File: generation/video_fetcher.py
"""
Script for fetching Anthony Fantano videof rom the YouTube API.
"""
from googleapiclient.discovery import build


API_SERVICE_NAME = "youtube"
API_VERSION = "v3"
MAX_PAGE_SIZE = 50


class VideoFetcher:
    """
    Class used to fetch YouTube videos from the YouTube Data API.
    """

    def __init__(self, api_key, debug=False):
        self.resource = build(
            API_SERVICE_NAME, API_VERSION, developerKey=api_key)
        self.debug = debug

    def fetch_num_playlist_videos(self, playlist_id):
        """ Function:   fetch_num_playlist_videos
            Parameters: playlist_id, the ID of the playlist to fetch from
            Returns:    int, the number of videos in this playlist.
        """
        request = self.resource.playlists().list(
            part="contentDetails",
            id=playlist_id
        )
        response = request.execute()
        return response["items"][0]["contentDetails"]["itemCount"]

    def fetch_playlist_page(self, playlist_id, page_token=None):
        """ Function:   fetch_playlist_page
            Parameters: playlist_id, the ID of the playlist to fetch from
                        page_token, string if specifying a page to fetch, None
                            if fetching the first page
            Returns:    (items, next_page_token), the items in the page and a
                        next_page_token if there is one
        """
        request = self.resource.playlistItems().list(
            part="contentDetails,snippet",
            playlistId=playlist_id,
            pageToken=page_token,
            maxResults=MAX_PAGE_SIZE
        )
        response = request.execute()
        items = response["items"]
        return (items, response.get("nextPageToken", None))

    def fetch_all_playlist_videos(self, playlist_id):
        """ Function:   fetch_all_playlist_videos
            Parameters: api_key, string, the YouTube API v3 key to use
                        playlist_id, the ID of the playlist to fetch from
            Returns: list of dict, list of metadata dictionaries for all videos
        """
        if self.debug:
            print("Fetching all theneedledrop videos...")

        num_videos = self.fetch_num_playlist_videos(playlist_id)

        # Request the first page.
        items, next_page_token = self.fetch_playlist_page(playlist_id)

        if self.debug:
            print(f"{len(items)}/{num_videos} videos fetched...", end='\r')

        # Request all subsequent pages and accumulate the results.
        while next_page_token is not None:
            new_items, next_page_token = self.fetch_playlist_page(playlist_id,
                                                                  next_page_token)
            items += new_items

            if self.debug:
                print(f"{len(items)}/{num_videos} videos fetched...",
                      end='\r')

        return items

    def fetch_video(self, playlist_id, video_id):
        """ Function:   fetch_video
            Parameters: playlistId, the ID of the channel's general playlist
                        videoId, the ID of the video
            Returns:    dict, video metadata
        """
        request = self.resource.playlistItems().list(
            part="contentDetails,snippet",
            playlistId=playlist_id,
            videoId=video_id
        )
        response = request.execute()

        if len(response["items"]) < 1:
            return None

        return response["items"][0]

File: src/videoFetcher.ts
import { youtube_v3 } from 'googleapis';
import { Either, EitherAsync, List, Maybe, Right } from 'purify-ts';
import { maybeOf } from './purifyUtils';
import logUpdate from 'log-update';

const MAX_PAGE_SIZE = 50;

// TODO: Handle exceptions from awaited promises.

export const getNumPlaylistVideos =
  (service: youtube_v3.Youtube) =>
  async (playlistId: string): Promise<Either<Error, number>> => {
    const response = await service.playlists.list(
      {
        id: [playlistId],
        part: ['contentDetails'],
      },
      {},
    );

    return maybeOf(response.data.items)
      .chain(List.head)
      .chainNullable((value) => value.contentDetails?.itemCount)
      .toEither(
        Error(
          `Unable to retrieve number of playlist videos from playlist ${playlistId}`,
        ),
      );
  };

type PlaylistPageWithNextPageToken = {
  items: youtube_v3.Schema$PlaylistItem[];
  maybeNextPageToken: Maybe<string>;
};

const getPlaylistPage =
  (service: youtube_v3.Youtube) =>
  async (
    playlistId: string,
    pageToken?: string,
  ): Promise<Either<Error, PlaylistPageWithNextPageToken>> => {
    const response = await service.playlistItems.list(
      {
        playlistId,
        pageToken,
        maxResults: MAX_PAGE_SIZE,
        part: ['contentDetails', 'snippet'],
      },
      {},
    );

    return maybeOf(response.data.items)
      .map((items) => ({
        items,
        maybeNextPageToken: maybeOf(response.data.nextPageToken),
      }))
      .toEither(
        Error(
          pageToken === undefined
            ? `Unable to fetch first page from playlist ${playlistId}`
            : `Unable to fetch page ${pageToken} from playlist ${playlistId}`,
        ),
      );
  };

export const getAllPlaylistVideos =
  (service: youtube_v3.Youtube) =>
  async (
    playlistId: string,
  ): Promise<Either<Error, youtube_v3.Schema$PlaylistItem[]>> => {
    // TODO: Right now logging is always on. It should be configurable.
    console.log('Fetching all theneedledrop videos...');

    const numVideos = (
      await getNumPlaylistVideos(service)(playlistId)
    ).orDefault(0);

    const logProgress = (acc: youtube_v3.Schema$PlaylistItem[]): void =>
      logUpdate(`${acc.length}/${numVideos} videos fetched...`);

    // Recursively fetch pages of playlist videos until there is no longer a
    // next page. That is indicated by there being no `nextPageToken` in a
    // response. Accumulator-based recursion is used here in order to enable a
    // better logging experience of the progress of fetching videos.
    const getNextPageRec = async (
      playlistId: string,
      acc: youtube_v3.Schema$PlaylistItem[],
      pageToken?: string,
    ): Promise<Either<Error, youtube_v3.Schema$PlaylistItem[]>> => {
      const eitherPageOrError = EitherAsync.fromPromise(() =>
        getPlaylistPage(service)(playlistId, pageToken),
      );

      return eitherPageOrError.chain((pageAndMaybeNextPageToken) => {
        const newAcc = acc.concat(pageAndMaybeNextPageToken.items);
        logProgress(newAcc);

        return pageAndMaybeNextPageToken.maybeNextPageToken.caseOf({
          Just: (nextPageToken) =>
            EitherAsync.fromPromise(() =>
              getNextPageRec(playlistId, newAcc, nextPageToken),
            ),
          Nothing: () => EitherAsync.liftEither(Right(newAcc)),
        });
      });
    };

    return getNextPageRec(playlistId, []);
  };

export const getVideo =
  (service: youtube_v3.Youtube) =>
  async (
    playlistId: string,
    videoId: string,
  ): Promise<Either<Error, youtube_v3.Schema$PlaylistItem>> => {
    const response = await service.playlistItems.list({
      part: ['contentDetails', 'snippet'],
      playlistId,
      videoId,
    });

    return maybeOf(response.data.items)
      .chain(List.head)
      .toEither(
        Error(`Unable to fetch video ${videoId} from playlist ${playlistId}.`),
      );
  };
